Documentation/manpages/initrd-scanmod.1.md
initrd-scanmod(1)

# NAME

initrd-scanmod - searches for kernel modules according to patterns

# SYNOPSIS

*initrd-scanmod* [options] [--] rules-file [rules-file ...]

# DESCRIPTION

The utility uses a set of rules to find and show kernel modules. It shows the
modules that satisfy all the rules.  This utility reads rules for filtering from
files. It can filter modules by the following criteria: *alias*, *author*,
*depends*, *description*, *filename*, *firmware*, *license*, *name* and *symbol*.

# OPTIONS

*-k, --set-version=*_VERSION_
	use _VERSION_ instead of ‘uname -r‘.

*-b, --base-dir=*_DIR_
	use _DIR_ as filesystem root for /lib/modules;

*-v, --verbose*
	print a message for each action.

*-V, --version*
	Show version of program and exit.

*-h, --help*
	Show this text and exit.

# RULES

The rules are in the format:

```
<keyword> <regular-expression>
```

All empty rules or starting with the ‘‘#’’ character are ignored. If the keyword
begins with the prefix ‘‘not-’’ then the rule will be inverted.

	*name*

	This allows you to filter modules by the kernel module name.

	*alias*

	This allows you to search the aliases list of a kernel module. For
	example:

	```
	alias ˆfs-
	```

	*depends*

	Modules depend on other modules. This keyword allows you to filter
	modules based on dependencies. Example:

	```
	depends ˆ(.*,)?af_alg(,.*)?$
	```

	*symbol*

	Also you can filter by module symbol names. This is a very shaky
	approach. The symbol names may change in the new kernel version and the
	filter will stop working. Some stable subsystems don’t change often.
	Example:

	```
	symbol ˆ(nfs_.*|svc_.*)$
	```

	or

	```
	symbol ˆregister_virtio_driver
	```

# AUTHOR

Written by Alexey Gladkov.

# BUGS

Report bugs to the authors.


NEWS.md
Version 2.39.0

- Feature plymouth:
  + Show splash in synchronous mode to avoid race with plymouth users.
- Feature fsck:
  + Fix typo in boot variable and fix the FASTBOOT parameter.
- Utilities:
  + initrd-scanmod: Add support for kmod > v30-25-g5c004af.

Version 2.38.0

- Runtime:
  + Rename monotonic-timestamp -> timestamp.
  + Use boottime instead of monotonic time.
- Feature dropbear:
  + Mount devpts if not mounted.
- Guess subsystem:
  + Add device suppliers to initrd.
- Remove unsupported bootloader feature.

Version 2.37.0

- Runtime:
  + Reduce few timeouts.
  + Get rid of localdev polling.
  + Convert rootdelay to service.
  + Drop extender scripts.
- Feature zfs:
  + Replace extender script by service.
  + Drop live-mode.

Version 2.36.0

- Runtime:
  + Reimplement ueventd, replace polld by uevent queue.
  + Create a shared libinitramfs library with shared code.
  + Use flock-based locking.
  + Rewrite console locking.
  + Add /dev to fstab and mount it as other filesystems.
  + Do not fail if required mountpoint does not exist.
- New feature:
  + Add debug-procacct. The feature is designed to debug the boot process inside
    the initramfs.
- Feature mdadm:
  + Boot from IMSM.
  + Put mdmon if external metadata detected.
- Feature system-glibc:
  + Do not hardcode min_uid/min_gid values.
- Utilities:
  + udev-rules: Add more udev rule checks.

Version 2.35.0

- Guess subsystem:
  + Do not fail if device filesystem type is swap.
- Misc:
  + Add udev rules parser. This is a new parser that allows you to search for
    syntactic and logical errors. It also allows you to more accurately obtain
    information from the rules.

Version 2.34.0

- Feature guestfs:
  + Add programs used by libguestfs.
  + Add raid modules and udev rules.
  + Config guestfs.mk.example: remove features already required by guestfs.
- Feature kickstart:
  + Do not show rsync progress on serial console.
  + Ask mdadm to create device nodes in /dev.
- Misc:
  + Update busybox 1.35.0.

Version 2.33.0

- Runtime:
  + Use pre/post handlers to extend cmdline service and remove service-specific
    handlers.
  + Replace mountsrc service by post-service handler.
  + Show block/net udev events if rdlog=console.
  + Add timestamps to polld messages.
- Feature zfs:
  + Add OpenZFS feature.
- Feature kickstart:
  + Add zfs support.
  + Add documentattion about zfs commands.
  + Do not try to get UUID for non-device mounts.
- Feature sshfsroot:
  + Add documentattion about root=/dev/sshfs.
- Features mdadm:
  + Use own copy of 63-md-raid-arrays.rules to fix regression with fedora boot
    on raid.
- Misc:
  + Sync gen_init_cpio with upstream.
  + Add documentation about module pattern sets.
  + Add documentation about 9P as root filesystem.

Version 2.32.1

- mk: do not expand functions out of '$(call ...)' context.
- testing: Add cmdline parameter to do not panic.

Version 2.32.0

- Runtime:
  + Reduce rootdelay period if all mountpoints are done, but init program is
    missing.
  + Show proper message if INIT not found.
- Feature luks:
  + Do not overwrite LUKS_CRYPTTAB.

Version 2.31.0

- Runtime:
  + Check more carefully for the presence of the INIT= inside new root partition.
- Feature kickstart:
  + Start all luks after partitioning.
  + Add simple reqpart that automatically creates partitions required by your
    hardware platform.
  + Add part --fstype=efi to create EFI partition with custom mountpoint.
  + Add --hibernation option to part/logvol/raid commands.  This option can be
    used to automatically determine the size of the swap partition big enough
    for hibernation.
  + Add support for fat/vfat filesystem.
  + Add support for zstd/lz4-compressed tarballs in liveimg.
- Feature qemu:
  + Try to add e1000e module.
- Feature network:
  + Fix permissions of /etc/resolv.conf
- Misc:
  + Relax check of /usr. This will allow to migrate the filesystem to /usr.

Version 2.30.0

- Feature kickstart:
  + Close luks partition if we need to change partition table.
  + Add information about what command is being executed.
- Feature rootfs:
  + Create fstab more carefully.
- Feature ucode:
  + Change path in the archive.
- Feature multipath:
  + Add more rules and utils for FC multipath.
- Drop buildinfo feature.
- Misc:
  + Detect separate /usr partition (merged-usr).
  + Generate wiki from Documentation.

Version 2.29.0

- Runtime:
  +  Remove shell service.
- Feature pipeline:
  + Give resume a chance to run.
  + Add wait-resume step.

Version 2.28.0

- Feature lkrg:
  + Support for LKRG v0.9.3-43-g49a3117.
- Feature pipeline:
  + Remove standalone daemon.
- Runtime:
  + Make log messages more readable.
  + Wait until the resume is processed.
- Misc:
  + Add more integration tests.

Version 2.27.2

- Feature luks:
  + Get rid of LUKS_ALL_CRYPTTAB. The LUKS_ALL_CRYPTTAB variable was needed
    until the autodetect worked.

Version 2.27.1

- Feature luks:
  + Fix typo and use keyfile if exists.
  + Add more documentation.

Version 2.27.0

- Feature luks:
  + Add crypttab support.
  + Try to remember the uuid of the luks device and automatically add it to crypttab.
- Feature kickstart:
  + Use /proc/devices to detect sd and virtblk block devices.
- Feature locales:
  + Read system-wide locales.
- Utilities:
  + initrd-ls: Fix infinite loop when unpacking zstd.
  + initrd-put: Add option to exclude files by pattern.
  + initrd-scanmod: module must satisfy all the rules from the ruleset.
  + mkinitrd-make-initrd: Fixed misprint, make-initrd can be found now.
  + make-initrd: Enforce absolute path in TMPDIR.
- Misc:
  + Сheck bzip2 library as the last chance for detection if there is no
    bzip2.pc.

Version 2.26.0

- Runtime:
  + Show a message if the root is not found for more than 15 seconds.
- New fearure:
  + Add locales feature. The feature adds locales and translations of utilities.
- Feature ucode:
  + Don't throw an error if no cpu updates are found for intel.
  + Fix firmware definition for amd cpu.

utils/initrd-scanmod/initrd-scanmod.c
#define _GNU_SOURCE
#include <sys/utsname.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>
#include <ctype.h>
#include <err.h>

#include "initrd-scanmod.h"
#include "config.h"

int verbose = 0;

const char short_opts[]         = "Vhb:k:v";
const struct option long_opts[] = {
	{ "set-version", required_argument, 0, 'k' },
	{ "base-dir", required_argument, 0, 'b' },
	{ "help", no_argument, NULL, 'h' },
	{ "verbose", no_argument, NULL, 'v' },
	{ "version", no_argument, NULL, 'V' },
	{ NULL, 0, NULL, 0 }
};

static void __attribute__((noreturn))
usage(const char *progname, int code)
{
	dprintf(STDOUT_FILENO,
	        "Usage: %s [options] [--] rules-file [rules-file ...]\n"
	        "\n"
	        "Options:\n"
	        " -k, --set-version=VERSION   use VERSION instead of `uname -r`;\n"
	        " -b, --base-dir=DIR          use DIR as filesystem root for /lib/modules;\n"
	        " -v, --verbose               print a message for each action;\n"
	        " -h, --help                  display this help and exit;\n"
	        " -V, --version               output version information and exit.\n"
	        "\n"
	        "Report bugs to authors.\n"
	        "\n",
	        progname);
	exit(code);
}

static inline void __attribute__((noreturn))
print_version_and_exit(const char *progname)
{
	dprintf(STDOUT_FILENO, "%s version %s\n", progname, PACKAGE_VERSION);
	dprintf(STDOUT_FILENO,
	        "Written by Alexey Gladkov.\n\n"
	        "Copyright (C) 2018  Alexey Gladkov <gladkov.alexey@gmail.com>\n"
	        "This is free software; see the source for copying conditions.  There is NO\n"
	        "warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n");
	exit(EXIT_SUCCESS);
}

int
main(int argc, char **argv)
{
	int c;
	struct utsname u;
	char *kerneldir      = NULL;
	const char *kversion = NULL;
	const char *basedir  = NULL;

	while ((c = getopt_long(argc, argv, short_opts, long_opts, NULL)) != EOF) {
		switch (c) {
			case 'k':
				kversion = optarg;
				break;
			case 'b':
				basedir = optarg;
				break;
			case 'v':
				verbose++;
				break;
			case 'h':
				usage(basename(argv[0]), EXIT_SUCCESS);
				break;
			case 'V':
				print_version_and_exit(basename(argv[0]));
				break;
			case '?':
				usage(basename(argv[0]), EXIT_FAILURE);
		}
	}

	if (argc == optind) {
		warnx("rules file required");
		usage(basename(argv[0]), EXIT_FAILURE);
	}

	if (!basedir)
		basedir = "";

	if (!kversion) {
		if (uname(&u) < 0)
			err(EXIT_FAILURE, "ERROR: uname()");
		kversion = u.release;
	}

	xasprintf(&kerneldir, "%s/lib/modules/%s", basedir, kversion);

	if (verbose)
		warnx("kernel directory: %s", kerneldir);

	parse_rules(argc - optind, argv + optind);
	find_modules(kerneldir);

	free_rules();
	xfree(kerneldir);

	return EXIT_SUCCESS;
}
